cmake/ChooseCudaCC.cmake

```diff
@@ -0,0 +1,89 @@
+#
+# after returning from this function, do not forget to call the following:
+#    set(RESULT_NAME ${RESULT_NAME} CACHE STRING "CUDA CC versions to compile")
+# replacing your own variable for RESULT_NAME
+#
+# We assume that MINCC default to 20
+# We assume that MINCUDAVERSION defaults to 7.0
+#
+function(chooseCudaCC RESULT_NAME MINCC MINCUDAVERSION)
```

Comment:
In order to deal with optional arguments we can use `cmake_parse_arguments`
```cmake
#
# Return the minimal set of supported Cuda CC 
#
# Usage:
#   chooseCudaCC(RESULT_NAME
#                [MIN_CUDA_VERSION X.Y]
#                [MIN_CC XX ])
#
# RESULT_NAME variable storing the list of supported CC.
# MIN_CUDA_VERSION the minimal supported version of cuda (e.g. 7.5, default 7.0).
# MIN_CC minimal supported Cuda CC by the project
#
function(chooseCudaCC RESULT_NAME)
  set(options "")
  set(oneValueArgs MIN_CUDA_VERSION MIN_CC)
  set(multiValueArgs "")
  cmake_parse_arguments(CHOOSE_CUDA "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )

  if(NOT DEFINED CHOOSE_CUDA_MIN_CC)
    set(CHOOSE_CUDA_MIN_CC 20)
  endif()

  if(NOT DEFINED CHOOSE_CUDA_MIN_CUDA_VERSION)
    set(CHOOSE_CUDA_MIN_CUDA_VERSION 7.0)
  endif()

  find_package(CUDA ${CHOOSE_CUDA_MIN_CUDA_VERSION} REQUIRED)

  if(NOT CUDA_FOUND)
    message(FATAL_ERROR "Could not find CUDA >= ${CHOOSE_CUDA_MIN_CUDA_VERSION}")
  endif()
```
like that you can call it:
```cmake
chooseCudaCC(RESULT_NAME MIN_CUDA_VERSION 7.0 MIN_CC 30)
# or
chooseCudaCC(RESULT_NAME)
# or
chooseCudaCC(RESULT_NAME MIN_CC 30)
# or
chooseCudaCC(RESULT_NAME  MIN_CC 30 MIN_CUDA_VERSION 7.0)
# or
chooseCudaCC(RESULT_NAME  MIN_CUDA_VERSION 7.0)
```

The only cons is that `cmake_parse_arguments` prepend a given suffix (e.g. `CHOOSE_CUDA`) to the variables, so you need to search and replace
```
MINCC --> CHOOSE_CUDA_MIN_CC
MINCUDAVERSION --> CHOOSE_CUDA_MIN_CUDA_VERSION
```
Obviously feel free to change `CHOOSE_CUDA` to anything else meaningful that does not bloat the code

Reply:
I've not understood the declaration yet.
Do I understand your example correctly, that I can write:
```
function(chooseCudaCC SUPPORTED_CC SUPPORTED_GENCODE_FLAGS)
  set(options "")
  set(oneValueArgs MIN_CUDA_VERSION MIN_CC)
  set(multiValueArgs "")
  cmake_parse_arguments(CHOOSE_CUDA "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )

  if(NOT DEFINED CHOOSE_CUDA_MIN_CC)
    set(CHOOSE_CUDA_MIN_CC 20)
  endif()

  if(NOT DEFINED CHOOSE_CUDA_MIN_CUDA_VERSION)
    set(CHOOSE_CUDA_MIN_CUDA_VERSION 7.0)
  endif()

  find_package(CUDA ${CHOOSE_CUDA_MIN_CUDA_VERSION} REQUIRED)

  if(NOT CUDA_FOUND)
    message(FATAL_ERROR "Could not find CUDA >= ${CHOOSE_CUDA_MIN_CUDA_VERSION}")
  endif()
```
What is the role of `options` if it's supposed to be done like this?

Reply:
so basically `SUPPORTED_CC` and `SUPPORTED_GENCODE_FLAGS` will catch the first two positional arguments passed to the function.
i.e. if u call
```cmake
chooseCudaCC(FOO BAR)
```
then `SUPPORTED_CC` will contain the value of FOO and so on.

Then the remaining arguments that are passed to the function will be parsed to look for named parameters and options. 

Basically you can tell cmake to look for: 

* options, are (optional) "flags" or switch that the user can pass, like e.g. `REQUIRED` in `find_package()`, if u set it in the call it is true, false otherwise. Here we don't have any, so we set to empty. If we wanted to add `REQUIRED` we would write `set(options REQUIRED)` and then after `cmake_parse_arguments` we can access to it as `CHOOSE_CUDA_REQUIRED `which will be true or false according to the call.

* oneValueArgs are named parameters that takes only one value, in our case we expect that the user could (optionally, because we use default values otherwise) pass MIN_CUDA_VERSION followed by a value and  MIN_CC by another value:
```cmake
chooseCudaCC(FOO BAR MIN_CUDA_VERSION 10 MIN_CC 62)
```
basically it is similar to some languages like python that allows to call a function with named parameters like 
```python
window.addNewControl(title="Title",
                     xPosition=20,
                     yPosition=50,
                     width=100,
                     height=50,
                     drawingNow=true)
```
where e.g. `MIN_CUDA_VERSION` is tje equivalent of specifying `xPosition`

* multiValueArgs is similar to the previous but it allows to pass multiple values, like a list.

The final catch is that `cmake_parse_arguments` will add the prefix you specify as its first argument to the named parameters it found, ie `MIN_CUDA_VERSION` becomes `CHOOSE_CUDA_MIN_CUDA_VERSION` because of the `CHOOSE_CUDA` that is passed. Which is completely arbitrary, it can be whatever prefix you like.

I know, not really easy to understand at first but that's cmake...

more info here https://cmake.org/cmake/help/latest/command/cmake_parse_arguments.html